control-plane/src/modules/deployment/deployment.ts

```diff
@@ -0,0 +1,180 @@
+import { and, eq, or, sql } from "drizzle-orm";
+import { LambdaProvider } from "./lambda";
+
+export type Deployment = {
+  id: string;
+  clusterId: string;
+  service: string;
+  packageUploadUrl: string;
+  definitionUploadUrl: string;
+  status: string;
+};
+
+export const s3AssetDetails = (
+  deployment: Deployment,
+): { S3Bucket: string; S3Key: string } => {
+  if (!UPLOAD_BUCKET) {
+    throw new Error("Upload bucket not configured");
+  }
+
+  return {
+    S3Bucket: UPLOAD_BUCKET,
+    S3Key: `${deployment.clusterId}/${deployment.service}/${deployment.id}-package`,
+  };
+};
+
+const lambdaProvider = new LambdaProvider();
+
+export const createDeployment = async ({
+  clusterId,
+  serviceName,
+}: {
+  clusterId: string;
+  serviceName: string;
+}): Promise<Deployment> => {
+  if (!UPLOAD_BUCKET) {
+    throw new Error("Upload bucket not configured");
+  }
+
+  const id = ulid();
+
+  const packageUploadUrl = await getPresignedURL(
+    UPLOAD_BUCKET,
+    clusterId,
+    serviceName,
+    `${id}-package`,
+  );
+  const definitionUploadUrl = await getPresignedURL(
+    UPLOAD_BUCKET,
+    clusterId,
+    serviceName,
+    `${id}-definition`,
+  );
+
+  // TODO: handle ID collision?
+  const deployment = await data.db
+    .insert(data.deployments)
+    .values([
+      {
+        id: id,
+        cluster_id: clusterId,
+        service: serviceName,
+        package_upload_path: packageUploadUrl,
+        definition_upload_path: definitionUploadUrl,
+        // Temporary, the expectation is that the deployment will be in the "uploading" while any async work is being done
+        status: "ready",
+      },
+    ])
+    .returning({
+      id: data.deployments.id,
+      clusterId: data.deployments.cluster_id,
+      service: data.deployments.service,
+      packageUploadUrl: data.deployments.package_upload_path,
+      definitionUploadUrl: data.deployments.definition_upload_path,
+      status: data.deployments.status,
+    });
+
+  return deployment[0];
+};
+
+export const getDeployment = async ({
+  clusterId,
+  serviceName,
+  id,
+}: {
+  clusterId: string;
+  serviceName: string;
+  id: string;
+}): Promise<Deployment> => {
+  const deployment = await data.db
+    .select({
+      id: data.deployments.id,
+      clusterId: data.deployments.cluster_id,
+      service: data.deployments.service,
+      packageUploadUrl: data.deployments.package_upload_path,
+      definitionUploadUrl: data.deployments.definition_upload_path,
+      status: data.deployments.status,
+    })
+    .from(data.deployments)
+    .where(
+      and(
+        eq(data.deployments.id, id),
+        eq(data.deployments.cluster_id, clusterId),
+        eq(data.deployments.service, serviceName),
+      ),
+    );
+
+  return deployment[0];
+};
+
+export const releaseDeployment = async (
+  deployment: Deployment,
+): Promise<Deployment> => {
+  // Check if any deployments have been released for this service
+  const releases = await data.db
```

Comment:
Could probably store a reference on `service` to the current deployment rather than scanning the whole table.